exchanges/okx/okx_websocket.go

```diff
@@ -0,0 +1,1848 @@
+		okxChannelMarkPriceCandle3Mutc, okxChannelMarkPriceCandle1Mutc, okxChannelMarkPriceCandle1Wutc, okxChannelMarkPriceCandle1Dutc,
+		okxChannelMarkPriceCandle2Dutc, okxChannelMarkPriceCandle3Dutc, okxChannelMarkPriceCandle5Dutc, okxChannelMarkPriceCandle12Hutc,
+		okxChannelMarkPriceCandle6Hutc:
+		return ok.wsHandleMarkPriceCandles(respRaw)
+	default:
+		ok.Websocket.DataHandler <- stream.UnhandledMessageWarning{Message: ok.Name + stream.UnhandledMessage + string(respRaw)}
+		return nil
+	}
+}
+
+// wsProcessIndexCandles processes index candlestic data
+func (ok *Okx) wsProcessIndexCandles(intermediate *WebsocketDataResponse) error {
+	if intermediate == nil {
+		return errNilArgument
+	}
+	var response WSCandlestickResponse
+	if len(intermediate.Data) == 0 {
+		return errNoCandlestickDataFound
+	}
+	pair, err := ok.GetPairFromInstrumentID(intermediate.Argument.InstrumentID)
+	if err != nil {
+		return err
+	}
+	var a asset.Item
+	a, _ = ok.GetAssetTypeFromInstrumentType(intermediate.Argument.InstrumentType)
+	candleInterval := strings.TrimPrefix(intermediate.Argument.Channel, candle)
+	for i := range response.Data {
+		candles, okay := (intermediate.Data[i]).([5]string)
+		if !okay {
+			return errIncompleteCandlestickData
+		}
+		timestamp, err := strconv.Atoi(candles[0])
+		if err != nil {
+			return err
+		}
+		candle := stream.KlineData{
+			Pair:      pair,
+			Exchange:  ok.Name,
+			Timestamp: time.UnixMilli(int64(timestamp)),
+			Interval:  candleInterval,
+			AssetType: a,
+		}
+		candle.OpenPrice, err = strconv.ParseFloat(candles[1], 64)
+		if err != nil {
+			return err
+		}
+		candle.HighPrice, err = strconv.ParseFloat(candles[2], 64)
+		if err != nil {
+			return err
+		}
+		candle.LowPrice, err = strconv.ParseFloat(candles[3], 64)
+		if err != nil {
+			return err
+		}
+		candle.ClosePrice, err = strconv.ParseFloat(candles[4], 64)
+		if err != nil {
+			return err
+		}
+		ok.Websocket.DataHandler <- candle
+	}
+	return nil
+}
+
+// wsProcessOrderBooks processes "snapshot" and "update" order book
+func (ok *Okx) wsProcessOrderBooks(data []byte) error {
+	var response WsOrderBook
+	var err error
+	err = json.Unmarshal(data, &response)
+	if err != nil {
+		return err
+	}
+	if response.Argument.Channel == okxChannelOrderBooks &&
+		!(response.Action == OkxOrderBookUpdate ||
+			response.Action == OkxOrderBookSnapshot) {
+		return errors.New("invalid order book action")
+	}
+	var pair currency.Pair
+	var a asset.Item
+	a, _ = ok.GetAssetTypeFromInstrumentType(response.Argument.InstrumentType)
+	if a == asset.Empty {
+		a = ok.GuessAssetTypeFromInstrumentID(response.Argument.InstrumentID)
+	}
+	pair, err = ok.GetPairFromInstrumentID(response.Argument.InstrumentID)
+	if err != nil {
+		pair.Delimiter = currency.DashDelimiter
+	}
+	for i := range response.Data {
+		if response.Action == OkxOrderBookSnapshot ||
+			response.Argument.Channel == okxChannelOrderBooks5 ||
```

Comment:
Remove `okxChannelOrderBooks5 ` as there is conflict between books5 which is an update and it will purge the current in memory book.